### PR TITLE
[OpenMP][MLIR] OMPEarlyOutliningPass removal

### PR DESCRIPTION
**Pull request by TIFitis**
This patch removes the OMPEarlyOutliningPass as it is no longer required. The implicit map operand capture has now been moved to the PFT lowering stage.

Depends on #67318.

### CODE REVIEW

**Comment by skatrak** [comment]
This change seems to break function filtering in certain cases. For example:
```fortran
! File: test.f90
program main
    !$omp target
    !$omp end target
    print *, 5
end program main
```

This results in a compiler crash around `convertDeclareTargetAttr()` in OpenMPToLLVMIRTranslation. It can be reproduced as follows: `flang-new -fc1 -fopenmp -fopenmp-is-target-device -emit-llvm -o - test.f90`.

The target outlining pass allowed the function filtering pass to assume that the only MLIR operations that remain represent either device code or wrapper functions around a single `omp.target` operation. Since those assumptions are now no longer met, the function filtering is failing.

**Comment by TIFitis** [comment]
@skatrak Thanks for pointing this out, I'm working on a fix now.

**Comment by agozillon** [comment]
The patch series LGTM so far, the only concern I have (which is very minor) is the map type and capture type we give the bounds information when we generate map_info's for them, but I think that's something we can adjust in future patches if we have any problems as I am not sure myself what they should be mapped as currently. Provided this patch series doesn't break any existing runtime tests we have I'm happy with it. 

I've also added one comment relating to manipulating the capture type of implicits based on how Clang handles it currently, we can add it in this patch series or later I don't mind too much, it just seemed like a good possible addition.

File: flang/include/flang/Optimizer/Transforms/Passes.td [hunk 1]
```diff
@@ -297,8 +297,8 @@ def PolymorphicOpConversion : Pass<"fir-polymorphic-op", "::mlir::func::FuncOp">
   let summary =
     "Simplify operations on polymorphic types";
   let description = [{
-    This pass breaks up the lowering of operations on polymorphic types by 
-    introducing an intermediate FIR level that simplifies code geneation. 
+    This pass breaks up the lowering of operations on polymorphic types by
+    introducing an intermediate FIR level that simplifies code geneation.
```
Comment (kiranchandramohan) [inline flang/include/flang/Optimizer/Transforms/Passes.td:301]: Nit: unrelated change.

File: flang/include/flang/Optimizer/Transforms/Passes.td [hunk 2]
```diff
@@ -16,7 +16,7 @@
 
 include "mlir/Pass/PassBase.td"
 
-class AbstractResultOptBase<string optExt, string operation> 
+class AbstractResultOptBase<string optExt, string operation>
```
Comment (kiranchandramohan) [inline flang/include/flang/Optimizer/Transforms/Passes.td:19]: Nit: unrelated change.

**Review by kiranchandramohan — COMMENTED** [review]
LG. Please wait for @agozillon.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 3]
```diff
@@ -2459,6 +2459,8 @@ convertDeclareTargetAttr(Operation *op, mlir::omp::DeclareTargetAttr attribute,
       if (declareType == omp::DeclareTargetDeviceType::host) {
         llvm::Function *llvmFunc =
             moduleTranslation.lookupFunction(funcOp.getName());
+        llvmFunc->replaceAllUsesWith(
+            llvm::UndefValue::get(llvmFunc->getType()));
```
Comment (TIFitis) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2463]: @skatrak I have add this line to replace the uses of the function that are to be removed with `undef` values. This should prevent the crash.

Let me know if this looks like an appropriate fix for the issue to you.

**Review by agozillon — APPROVED** [review]
LGTM, although, please do wait on @skatrak input on the fix you've pinged him in, it looks reasonable to me, but he will understand the implications of the change a lot better than I do. And thank you for your hard work on this amazing series of PRs.

**Comment by TIFitis** [comment]
Pushed commit 63752399f8323460f685c8431aae7269ea3f27a2.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp (hunk 3 above)
Comment (skatrak) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2463]: This looks like a good solution to the problem. Thanks!
